### ISSUE TITLE
Tests is failing on Alpine with V8 >= 5.7.434

### ISSUE THREAD

**Issue by AlexMasterov** [pinepain/php-v8#16]
[build.log](https://gist.github.com/anonymous/844f7fa4458c5afd5db9dce5923659c4) (master branch)

It is possible that I have a problem with V8, but `V8Js` ext works fine.

### How to test
[Dockerfile V8 5.7+ on Alpine](https://github.com/AlexMasterov/dockerfiles/blob/alpine/v8-gn/alpine-v8-lib/Dockerfile)
```sh
git clone -b alpine/v8-gn https://github.com/AlexMasterov/dockerfiles.git
cd dockerfiles/alpine-v8-lib
```
or you can just take the result: [V8 libs](https://www.dropbox.com/s/t42mkk5wrfhcoxd/alpine-v8-lib_5.7.434.tar.gz?dl=0) `/usr/local/v8`
```sh
V8_VERSION="5.7.434" \
V8_LIB_SOURCE="https://www.dropbox.com/s/t42mkk5wrfhcoxd/alpine-v8-lib_${V8_VERSION}.tar.gz"
curl -fSL --connect-timeout 30 ${V8_LIB_SOURCE} | tar xz -C
```
or you can just take my the minimum [Dockerfile php 7.0.14](https://gist.github.com/anonymous/205719afd1cebd22e48f7bed6b92f0ef).

(RU)
Если что, у меня также есть Skype и Telegram. 
Готов в меру сил оказать содействие в решение.

**Comment by pinepain** [comment]
Hi! Thank you for reporting. Sorry for late reply, life was too busy past week.

Can you please post what exactly fails in that tests: I have [test-report.sh](https://github.com/pinepain/php-v8/blob/master/test-report.sh) which I use to show output of failed tests: `make test | tee result.txt && test-report.sh` should do the job.

From your Docker file [alpine-libv8](https://github.com/AlexMasterov/dockerfiles/blob/master/alpine-libv8/Dockerfile), [it looks like](https://github.com/AlexMasterov/dockerfiles/blob/d41d02c3242e330e015d0c75c4a7c1da36667606/alpine-libv8/Dockerfile#L73-L75) that you don't copy `icudtl.dat`, which php-v8 expects to have (and from build flags it looks like that v8 build with icu, which is default behavior), so I guess this is the cause of an issue with failed tests. As tests that doesn't init V8 passes, it also makes me thing that missed ICU data file is the cause of tests failure. 

You can give a look how I package libv8 for [ubuntu](https://github.com/pinepain/ppa-packaging/blob/d3263c8ffe81dd5529f1dc088a46e8ebc47fb2ba/packaging/libv8/debian/rules#L79-L90) and [macOS (homebrew)](https://github.com/pinepain/homebrew-devtools/blob/3130b193746a5ed64d5a293a3abfbdb5eb91d44e/Formula/v8%405.7.rb#L71-L74).

In latest changes I make php-v8 be quite demanding to libv8 and it expects vanilla build, which has own ICU. While it's a nice thing, as before I faced with a minor problems as icu versions was different across ubuntu versions and on macOS, it also has a drawback for our case - it has it data as a standalone file rather then built-in. As v8js expects non-icu libv8 and it doesn't initialize it, it should works just fine, but only with non-icu libv8 builds, thought, I'm not quite sure about that, so to validate my assumption, you can run some of [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat#Using_locales) examples with v8js.

(RU)
Огромное спасибо! Отписал на почтовый ящик указанный в гитхаб профиле.

**Comment by AlexMasterov** [comment]
Thanks for the reply.

[gcc](https://pkgs.alpinelinux.org/package/edge/main/x86_64/gcc) 6.3.0-r1
[icu-dev](https://pkgs.alpinelinux.org/package/edge/main/x86_64/icu-dev) 58.2-r1
[alpine-libv8: Add icudtl.dat in shared libraries](https://github.com/AlexMasterov/dockerfiles/commit/a48a2f5830c42878655c66dd23a85d944adf24b8)

[test-result.log](https://gist.github.com/anonymous/8b824d63ef4cd902a0fefb51efcea8d5)
[test-report.log](https://gist.github.com/anonymous/675885df71053e02276afd2691636fcf)

(RU)
Постараюсь позже ответить подробней на почту. Флаги протестирую.

**Comment by pinepain** [comment]
Hi, thanks for detailed logs!

> 002+ # Fatal error in ../../src/snapshot/natives-external.cc, line 122
003+ # Check failed: holder_.

Add `v8_use_external_startup_data=0` to build flags

TL;DR

It looks like you have startup data as external file, php-v8 expect libv8 built with `v8_use_external_startup_data=0`. At this time I don't detect external snapshot in build script as having built-in startup data offers better performance and I can't easily come up with a generic use case when end-users will alter external startup data, especially from PHP userland. V8js, is more lenient to libv8 options and [it try to detect](https://github.com/phpv8/v8js/blob/6a53d03d193bd8f294f591c9adf801e820cc7576/config.m4#L154-L174) such variations. While the same approach for blobs detection could be added to php-v8, I'm not fond of supporting of libv8 builds variability at this time, you'd better just add `v8_use_external_startup_data=0` to build flags and the issue should be resolved for this particular case.

**Comment by AlexMasterov** [comment]
[alpine-libv8: Add "v8_use_external_startup_data=false" to build flags](https://github.com/AlexMasterov/dockerfiles/commit/0d03c46753d47be3951df8e70861049f4d121d29)

New report looks better.

[test-result.log](https://gist.github.com/anonymous/3b33717605ad82adb135bf51610c3d6c)
[test-report.log](https://gist.github.com/anonymous/e495e9ae5d4e456aa11d941ffd064ed6)

```sh
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
V8\DateObject [tests/V8DateObject.phpt]
V8\Int32Value [tests/V8Int32Value.phpt]
V8\IntegerValue [tests/V8IntegerValue.phpt]
V8\Uint32Value [tests/V8Uint32Value.phpt]
```
[`V8DateObject.phpt`](https://github.com/pinepain/php-v8/blob/fcc9eb1808831338dc6477eda03f828a57062898/tests/V8DateObject.phpt)
`date.timezone` is empty in php.ini; this is the only problem? v8js ext. has something similar:

```sh
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Test V8::registerExtension() : Register extension with errors [tests/extensions_error.phpt]
Test V8::executeString() : Check timezone handling [tests/timezones.phpt]

**Comment by pinepain** [comment]
Great! Thanks! Awesome work!

Timezone problem comes from environment timezone, last time i was looking into this specific proble, v8 respect `TZ` env variable, so having it preset to some standard value should fix test result inconsistency. I'll give it a look later today (as well as fixing float/int argument issue).

**Comment by pinepain** [comment]
I fixed `Int*Value` tests, they should be fine now.

As to `DateObject` test, under the hood it also tests timezone change at runtime. From your test output I can assume that somehow internal date cache is not reset during `V8\DateObject::DateTimeConfigurationChangeNotification()` call. But it may be that Alpine handles timezone differently from other distros or it's some issue with it libc++ or in ICU.

**Comment by AlexMasterov** [comment]
This report from the latest commit [`b33c148`](https://github.com/pinepain/php-v8/commit/b33c148aa54634ae8205154d08ca5ed27c0d248a).
No change. Hm.

[test-result.log](https://gist.github.com/anonymous/46ef439e798af8e9c51bab58cbdd8eef)
[test-report.log](https://gist.github.com/anonymous/56c27b3b65e44a689c38d999557ae8fd)

**Comment by pinepain** [comment]
Hm, my fault, "fixed" nonexistent issue then. 

```
double: NAN
128+ V8\IntegerValue->Value(): int(0)
128- TypeError: Argument 2 passed to V8\IntegerValue::__construct() must be of the type integer, float given\n
```

It seems to be that for some reason NAN is integer zero. I'm a bit surprised. What libc++ does Alpine has and which one do you use to build php-v8? Can you check what NAN is in Alpine?

**Comment by AlexMasterov** [comment]
[musl-libc](http://www.musl-libc.org/faq.html)

> Can you check what NAN is in Alpine?

`float(NAN)`
The same output as here: [is-nan()](http://php.net/is-nan).

**Comment by pinepain** [comment]
It seems to be a bug, https://bugs.php.net/bug.php?id=73954.

I guess there is nothing to do in advance with both `NAN` and timezone changing issues at this time as they both come from external tools. For now you may locally add `--XFAIL--` to known to fail tests or ignore that failures. I'll try to keep you  updated about a progress, if any.

**Comment by AlexMasterov** [comment]
@pinepain 
Thank you. I will support your report about this bug.

I think this issue can be closed.

**btw**, `is_nan(NAN)` returns FALSE on Alpine (musl) :smiley::+1:
